docs: add tutorial for reading plans (#354)
substrait-io/substrait: site/docs/tutorial/sql_to_substrait.md

# SQL to Substrait tutorial

To provide an introduction to Substrait for those already familiar with SQL,
this tutorial walks through the basics of Substrait by translating basic SQL
queries. We'll present the Substrait in JSON form so it is (somewhat) readable
for newcomers.

!!! note
    Substrait is currently only defined with Protobuf. The JSON provided here is the
    Protobuf JSON output, but it is *not* the official Substrait text format. Eventually,
    Substrait will define it's own human-readable text format, but for now this
    tutorial will make due with what Protobuf provides.

Substrait is designed to communicate plans (mostly logical plans). Those plans
contain types, schemas, expressions, extensions, and relations. We'll look at 

> them in that order, going from simplest to most complex until we can construct
> full plans.

Reviewer: nit: Suggestion

```suggestion
them in that order, going from basic to the most complex until we can construct
complete plans.
```

> ![](substrait_components.svg)

Reviewer 2: One minor nit about this image.  It shows four categories: types, relations, expressions, and functions.

I don't normally think of "functions" on the same level as the other three. Although, I'm also not sure exactly what "functions" is referring to, and maybe this is ok.

An expression can have a function invocation.  For example, a `ScalarFunction` message or a `WindowFunction` message.  But these are just a type of expression and so don't really deserve their own box.

On the other hand, a plan has a list of functions in the extension set (e.g. defines the URL/name/anchor).  If you're referring to this I suppose it's ok but in my mental picture I think of things like "read types" and "advanced extensions" at the same level of significance (e.g. not quite as significant as types/relations/expressions).

Anyways, I'm not sure anything here needs to change, it just seemed slightly off to me and I wanted to mention it.  If you're happy with it then we can stick with it as is.

Author: Yeah, this diagram is adapted from one of Jacques presentations. It's not really meant to accurately describe the relationships between each entity (except for the fact that plans are the top-level thing). It just gives an idea of what kinds of things are in plans.

I think it's fine for now, but if someone wants to design a better alternative, I have no qualms with replacing this.

> This tutorial won't cover all the details of each piece, but will give you an

Reviewer: nit: maybe? 

```suggestion
This tutorial won't cover all the details of each piece, but it will give you an
```

      }
    },
    "nullability": "NULLABILITY_REQUIRED"
  }
}
```

Schemas of tables can be represented with a `NamedStruct` message, which is the
combination of a struct type containing all the columns and a list of column
names. For the `orders` table, this will look like:

```json
{
  "names": [
    "product_id",
    "quantity",
    "order_date",
    "price"
  ],
  "struct": {
    "types": [
      {
        "i64": {
          "nullability": "NULLABILITY_REQUIRED"
        }
      },
      {
        "i32": {
          "nullability": "NULLABILITY_REQUIRED"
        }
      },
      {
        "date": {
          "nullability": "NULLABILITY_REQUIRED"
        }
      },
      {
        "decimal": {
          "scale": 10,
          "precision": 2,
          "nullability": "NULLABILITY_NULLABLE"
        }
      }
    ],
    "nullability": "NULLABILITY_REQUIRED"
  }
}
```

Here, `names` is the names of all fields. In nested schemas, this includes the
names of subfields in depth-first order. So for the `products` table, the 
`details` struct field will be included as well as the two subfields (`manufacturer`
and `year_created`) right after. And because it's depth first, these subfields

> appear before `product_name`. The full schema looks like:

Author: Is this correct? There is no subfield for list or map types? (categories here is a list of strings)

Reviewer 2: I think so but I'm not certain.  I don't think a list column has names (even if it is a fixed size list) for its children.  In a map column the names are part of the data and not the field.

                "enum": {
                  "specified": "NAN_IS_NOT_NAN"
                }
              },
              {
                "value": {
                  "literal": {
                    "string": "Computers"
                  }
                }
              },
              {
                "value": {
                  "selection": {
                    "directReference": {
                      "structField": {
                        "field": 1
                      }
                    },
                    "rootReference": {}
                  }
                }
              }
            ]
          }
        }
      }
    ]
  }
}
```

## Relations

In most SQL engines, a logical or physical plan is represented as a tree of 
nodes, such as filter, project, scan, or join. The left diagram below may be a 
familiar representation of our plan, where nodes feed data into each other
moving from left to right. In Substrait, each of these nodes is a **Relation**.

![](plan_tree_versus_expression.svg)

A relation that takes another relation as input will contain (or refer to) that
relation. This is usually a field called `input`, but for complex cases that

> have multiple inputs such as joins may have different keys. In JSON, the rough 

Reviewer: Just want to know what you meant by this statement :)

> have multiple inputs such as joins may have different keys. In JSON, the rough 

Author: Yeah that is complex. I just meant that it's usually called `input`, but in the join case it's called `left` and `right`, so it's not always the same field for the inputs.

      "scalarFunction": {
        "functionReference": 2,
        "outputType": {
          "bool": {
            "nullability": "NULLABILITY_NULLABLE"
          }
        },
        "arguments": [
          {
            "value": {
              "selection": {
                "directReference": {
                  "structField": {
                    "field": 0
                  }
                }
              }
            }
          },
          {
            "value": {
              "selection": {
                "directReference": {
                  "structField": {
                    "field": 4
                  }
                }
              }
            }
          }
        ]
      }
    }
  }
}
```

The final aggregation requires two things, other than the input. First is the
groupings. We'll use a single grouping expression containing the references to
the fields `product_name` and `product_id`. (Multiple grouping expressions can
be used to do cube aggregations.) These field references are indices, and we'll
discuss later in [](#field-indices) where these come from.

For `measures`, we'll need to define `sum(quantity * price) as sales`. Substrait

> is stricter about data types, and quantity is a integer while price is a

Reviewer: ```suggestion
is stricter about data types, and quantity is an integer while price is a
```

                      ]
                    }
                  }
                }
              }
            ]
          }
        }
      }
    ]
  }
}
```

## Field indices

So far, we have glossed over the field indices. Now that we've built up each
of the relations, it will be a bit easier to explain them.

Instead of referencing fields by name, in Substrait we always reference them by
index. Nested fields are part of the index sequence, put in depth-first order.
For example, as shown in the diagram below, in the `products` table read relation
the `manufacturer` and `year_created` fields come right after their parent field,
`details`.

One complication is that whenever we reference by field index, those field
indices are dependent on the context of the relation. The exact field indices
can change from relation to relation. For example, a join relations field 
indices are based on the combination of the left inputs fields and then the 
right input fields. And, while we don't have one in our example, a Project relation
adds new fields, which are visible to it's downstream relations. The diagram
below shoes the mapping of field indices within each relation and how each of

> the field references show up in each relations properties.

Author: Have I described the rules adequately / accurately? Is there anything important I've missed?

Reviewer 2: > those field indices are dependent on the context of the relation

> a join relations field indices

If I were to be pedantic I would say "the context of the expression".  A join relation does not have field indices.  A join relation has a "join expression" and the context of references in that expression is the output schema of the join relation.

An interesting case to consider (and I have a really old PR I need to get merged clarifying this :grimacing: https://github.com/substrait-io/substrait/pull/271/files) is the read relation.

A read relation has a filter expression and a projection (which isn't an expression exactly, but a masked expression).  A read relation effectively has _two_ schemas (maybe even more if we want to consider evolution but we won't go there :laughing: ).  There is the schema of the data on disk (the direct schema) and the output schema of the read relation (the direct schema after the projection mask has been applied to it).  So we have to be specific when talking about the filter expression which schema we are actually binding against.

Author: Thanks. It is complex, but I found that `substrait-validator` provides a really helpful UI for showing those schemas and so I've referenced that tool in the tutorial.

![](field_indices_layout.svg)

## Column selection and emit

Looking at the relations so far, you might be wondering: when do we select the
output columns?

You might be tempted to add a [Project relation](/relations/logical_relations/#project-operation) 
at the end to select a subset of columns. However, the project relation only
adds columns; it is not responsible for subsetting or reordering columns.

<!-- TODO: Is this true? -->
Instead, any relation can reorder or subset columns through the `emit` property.
By default, it is set to `direct`, which outputs all columns "as is". But it can

> also be specified as a sequence of field indices.

Author: Just want to make sure? Is this actually true? Is **any** relation able to subset or reorder columns?

Reviewer 2: Yes.  This is quite different from some existing tooling in this area (e.g. Acero, dplyr, ibis) and so if you are familiar with those then it can be a bit odd.  It has been a consistent source of confusion :).

It is a very "logical" concept and allows for writing compact plans.  One way to think about it is to recognize that just about any relation might be the terminal point for a particular column:

 * You might load a column purely for the sake of applying a filter and can drop it after the filter
 * You might load a column purely to use as a join key and drop it after the join
 * Etc.

So, rather than be forced to write a plan that is full of project nodes you can instead just emit at any point to drop a column once you are finished with it.

A corollary confusion is that the project node, by itself, cannot actually drop or reorder columns.  The project node, in Substrait, only adds columns to the end.  The only way to drop columns in a project node is with an emit.

This means that `SELECT expr1, expr2, expr3` typically gets represented as:

`project(expr1, expr2, expr3):emit=[len(input)+0, len(input)+1, len(input)+2]`

so far there's been no place to tell a consumer what those function reference
IDs mean or which extensions we are using. That information, goes at the plan
level.

The overall layout for a plan is

```json
{
  "extensionUris": [ ... ],
  "extensions": [ ... ],
  "relations": [
    {
      "root": {
        "names": [
          "product_name",
          "product_id",
          "sales"
        ],
        "input": { ... }
      }
    }
  ]
}
```

The `relations` field is a list of Root relations, which provides the final
column names for our query. The input to this relation is our aggregate relation
(which contains all the other relations as children).

For extensions, we need to provide `extensionUris` with the locations of the 
YAML files we used and `extensions` with the list of functions we used and which
extension they come from.

In our query, we used:

 * `index_in` (0), from `functions_string.yaml`,
 * `is_null` (1), from `functions_comparison.yaml`,
 * `equal` (2), from `functions_comparison.yaml`,
 * `sum` (3), from `functions_arithmetic_decimal.yaml`,
 * `multiply` (4), from `functions_arithmetic_decimal.yaml`.

So first we can create the three extension uris:

> <!-- TODO: what is the URI format? -->

Author: All the plans I have seen from ibis-substrait omits the URI, but that seems... wrong? Have we established the pattern for these URIs?

Reviewer 2: > Have we established the pattern for these URIs?

Not quite, but the time is ripe to fix this.  All produers I've seen (Isthmus & DuckDb) operate the same as Ibis.  Originally the contention was between:

 * Brittle URIs that are tied to a specific version of an extension function but the content is fixed (e.g. https://github.com/substrait-io/substrait/blob/v0.18.0/extensions/functions_arithmetic.yaml)
 * Stable URIs that are not tied to a specific version and the content may vary (e.g. https://github.com/substrait-io/substrait/blob/main/extensions/functions_arithmetic.yaml)

Based on the conversation in https://github.com/substrait-io/substrait/issues/274 I believe we were leaning towards the latter (stable URI, potentially changing content).

Note that, in Acero, we treat the URI `/` (or the empty string) as a special case where we fallback to name-only lookup and look for a function with the given name, regardless of what URI it maps to (and if there are multiple I think we pick the first we find).  This allows us to consume plans from producers today that omit the URI.

# SQL to Substrait tutorial

To provide an introduction to Substrait for those already familiar with SQL,
this tutorial walks through the basics of Substrait by translating basic SQL
queries. We'll present the Substrait in JSON form so it is (somewhat) readable
for newcomers.

> Substrait can be a type, an expression, a relation, or a plan. We'll look at them in

Reviewer 2: Substrait is a spec for plans.   A plan is made up of those building blocks but I'm not sure it is valid to say "Substrait can be a type, an expression, a relation, or a plan".

> those order, going from simplest to most complex.

Reviewer 2: ```suggestion
that order, going from simplest to most complex.
```

## Types and Schemas

<!-- TODO -->

## Expressions

A basic building block of queries are expressions. Let's start simple with the
SQL expression:

```sql
a > 10
```

An expression can be one of several things. `a` itself will be translated to
a field reference:

```json
{
  "selection": {
    "directReference": {
      "structField": {

>         "field": 1

Reviewer 2: Might be worth noting that Substrait refers to fields numerically based on some kind of schema / field order (depends on the operator).  In SQL field references are always named references.

Author: Good point. I'll make sure to call that out. I'll need to look a little deeper to, since the natural question is "relative to *which* schema?"

      }
    },
    "rootReference": {}
  }
}
```

!!! note
    Protobuf may not serialize fields with integer type and value 0, since 0 is the default.
    So if you instead saw `"structField": {}`, know that is is equivalent to 
    `"structField": { "field": 0 }`.

`10` will be translated to a literal expression:

```json
{
  "literal": {

>     "i8": 10,

Reviewer 2: Why is it an `i8`?

Author: :shrug: that's what Ibis emitted and it seemed fine to me. Though I suppose usually when people represent an integer it's an i32 or i64, so maybe it would be less weird if it was either of those.

I should probably explain this type to the reader (perhaps that's what you are getting at). One question I have: To what extent should a Substrait producer strive to keep the operands of the same type? If `a` in `a > 10` is an `i64` is it okay if I label `10` as a `i32`?

    "nullable": true
  }
}
```

Then full expression `a > 10` will be a scalar function expression, and will 
contain the expressions for `a` and `10` in it's `arguments` field:

```json
{
  "scalarFunction": {
    "functionReference": 2,
    "outputType": {
      "bool": {
        "nullability": "NULLABILITY_NULLABLE"
      }
    },
    "arguments": [
      {
        "value": {
          "selection": {
            "directReference": {
              "structField": {
                "field": 1
              }
            },
            "rootReference": {}
          }
        }
      },
      {
        "value": {
          "literal": {
            "i8": 10,
            "nullable": true
          }
        }
      }
    ]
  }
}
```

The name of the function is defined by the `functionReference`, which points to
the index of the extension function to use. This is serialized separately, which
you will see later in serialized plans.

```json
{
  "extensionFunction": {
    "extensionUriReference": 1,
    "functionAnchor": 2,

>     "name": "gt"

Reviewer 2: It might be a good idea to explicitly mention here that `gt` is equivalent to `>` in your above SQL expression.  Not everyone will be hard-wired to see "gt" and think "greater than".

Author: Good point. I'll also point out where the YAML definition is.

  }
}
```

## Relations

> An logical or physical plan is typically represented as a tree of nodes, such

Reviewer 2: ```suggestion
A logical or physical plan is typically represented as a tree of nodes, such
```
Do we want to introduce or mention "logical" and "physical"?  Won't these terms confuse a reader at this point?

Author: It's not clear here, but I was speaking of plans in general, not Substrait plans specifically. I assume that anyone coming from the SQL world has seen some form of logical / physical plan (at the very least, has run `EXPLAIN ...` at some point and tried to read the output). I'll make this more specific so that's clear.

as filter, project, scan, or join. Each of these nodes is a **Relation** in 

> Substrait.

Reviewer 2: I feel like, as a naive reader, I might be confused here.  What is a relation exactly?  A part of a tree?  Why is it a tree?  What does this tree represent?  How is a relation different from a function?

For example, to read a named table, we use a `read` relation. This looks like:

```json
{
  "read": {
    "baseSchema": {
      "names": [
        "b",
        "a"
      ],
      "struct": {
        "types": [
          {
            "string": {
              "nullability": "NULLABILITY_NULLABLE"
            }
          },
          {
            "i64": {
              "nullability": "NULLABILITY_NULLABLE"
            }
          }
        ],
        "nullability": "NULLABILITY_REQUIRED"
      }
    },
    "namedTable": {
      "names": [
        "my_database",
        "my_table"
      ]
    }
  }
}
```

Read relations provide a schema in `baseSchema` as discussed in the prior section.

> The type of read is a named table, so the `namedTable` field is present with 

Reviewer 2: The type of read?  There is more than one? <-- Something a naive reader might be thinking here.

`names` containing the list of name segments (`my_database.my_table`).

Read relations are leaf nodes, so other relations will have them nested within.

> For example, a filter relation of (`a > 10`) will look like:

Reviewer 2: Maybe...

Read relations are leaf nodes.  Leaf nodes don't depend on any other node for data and usually represent a source of data in our plan.  Leaf nodes are then typically used as input for other nodes that manipulate the data.

Reviewer 3: Do you think we need to refer to the docs https://substrait.io/relations/logical_relations/ where has a table shows the input and output number of each relation? Like read relation here require 0 input so that's why it's a leaf relation.

```json
{
  "filter": {
    "input": {
      "read": ...
      }
    },
    "condition": ...
  }
}
```

## Plans

> First, let's look at reading a table:

Reviewer 2: I feel like the first paragraph should explain what a plan is and not be giving an example.

```sql
SELECT * FROM my_table;
```

A plan has a list of relations, and each top one is a root relation. The root

> relation just has one input relation and then the names of the fields.

Reviewer 2: These are facts, and they are true, but I'm not sure what the story you are trying to tell is.  Perhaps something like...

Many plans are represented by a single tree of relations.  The root of the tree is the output of the plan and the leaves in the tree are the sources of data.

Howevr, some plans actually contain multiple trees.  One reason for this is when a plan has a common subtree that occurs multiple times throughout the plan.  That subtree can be pulled out and described on its own and then can be referenced in other trees.  This is why a Substrait plan is actually a list of relations.  Only one of these is a "root relation" and it forms the root of the plan.

      "extensionUriReference": 0,
      "functionAnchor": 0,
      "name": "index_in"
    }
  },
  {
    "extensionFunction": {
      "extensionUriReference": 1,
      "functionAnchor": 1,
      "name": "is_null"
    }
  },
  {
    "extensionFunction": {
      "extensionUriReference": 1,
      "functionAnchor": 2,
      "name": "equal"
    }
  },
  {
    "extensionFunction": {
      "extensionUriReference": 2,
      "functionAnchor": 3,
      "name": "sum"
    }
  },
  {
    "extensionFunction": {
      "extensionUriReference": 2,
      "functionAnchor": 4,
      "name": "multiply"
    }
  }
]
```

Our plan outputted in full is: [final_plan.json](./final_plan.json).

## Next steps

**Validate and introspect plans using 
[substrait-validator](https://github.com/substrait-io/substrait-validator)**.
Amongst other things, this tool can show what the current schema and column
indices are at each point in the plan. Try downloading the final plan JSON above
and generating an HTML report on the plan with:

```sh
substrait-validator final_plan.json --out-file output.html
```

**Generate more Substrait plans from SQL using DuckDB's compiler**. See the 
documentation on

> [the DuckDB Substrait extension](https://duckdb.org/docs/extensions/substrait).

Author: I'd love to point readers to one of the tools that currently can be used to generate Substrait from SQL.

DuckDB looked easy to install, but not sure if they want that attention yet. I noticed the instructions on the website had some typos and the provided example errors when run in Python. From the github repo, they aren't accepting bug reports except from those who have pre-arranged agreements.

I didn't see a user-facing way to install Isthmus (just compilation instructions), but correct me if I'm wrong.

Reviewer: Frankly I also think the existing docs could be improved further. I am not an expert in Graal based builds, but I also think it would be great to document the build, install and usage of isthmus with little more detail. I guess non-Java developers would also use this.

ambiguities.

## Problem Set up

To learn Substrait, we'll build up to a specific query. We'll be using the
tables:

```sql
CREATE TABLE orders (
  product_id: i64 NOT NULL,
  quantity: i32 NOT NULL,
  order_date: date NOT NULL,
  price: decimal(10, 2)
);
```

```sql
CREATE TABLE products (
  product_id: i64 NOT NULL,
  categories: list<string NOT NULL> NOT NULL,
  details: struct<manufacturer: string, year_created: int32>,
  product_name: string
);
```

This `orders` table represents events where products were sold, recording how
many (`quantity`) and at what price (`price`). The `products` table provides
details for each product, with `product_id` as the primary key.

And we'll try to create the query:

```sql
SELECT
  product_name,
  product_id,
  sum(quantity * price) as sales
FROM
  orders
INNER JOIN
  products
ON
  orders.product_id = products.product_id
WHERE
  -- categories does not contain "Computers"
  INDEX_IN("Computers", categories) IS NULL
GROUP BY
  product_name,
  product_id
```

The query asked the question: For products that aren't in the `"Computer"`
category, how much has each product generated in sales?

> However, Substrait doesn't correspond to SQL as much as it does to logical plans.

Reviewer 2: ```suggestion
However, Substrait doesn't correspond to SQL as much as it does to query plans.
```

Minor nit: Substrait could be used to represent both logical and physical plans.  This sentence _could_ potentially lead someone to thinking it is only for logical plans.

So to be less ambiguous, the plan we are aiming for looks like:

```
|-+ Aggregate({sales = sum(quantity_price)}, group_by=(product_name, product_id))
  |-+ InnerJoin(on=orders.product_id = products.product_id)
    |- ReadTable(orders)
    |-+ Filter(INDEX_IN("Computers", categories) IS NULL)
      |- ReadTable(products)
```

## Types and Schemas

As part of the Substrait plan, we'll need to embed the data types of the input
tables. In Substrait, each type is a distinct message, which at a minimum 
contains a field for nullability. For example, a string field looks like:

```json
{
  "string": {
    "nullability": "NULLABILITY_NULLABLE"
  }
}
```

Nullability is an enum not a boolean, since Substrait allows 
`NULLABILITY_UNSPECIFIED` as an option, in addition to `NULLABILITY_NULLABLE`

> (nullable) and `NULLABILITY_REQUIRED` (not nullable).

Reviewer 2: Eh...generally `_UNSPECIFIED` is an invalid value for an enum and will be rejected by all plans and consumers.  Pretty much every enum has an `_UNSPECIFIED` variant, even when it doesn't make sense (e.g. `JOIN_TYPE_UNSPECIFIED`).  This is done whenever an enum doesn't have a natural default value because otherwise there is no way in protobuf to distinguish between "use the default" and "this is missing".  So I'm not entirely sure why nullability is a boolean.  @jacques-n is `NULLABILITY_UNSPECIFIED` meant to be a legitimate thing?  I would think, if someone doesn't know, they would just pick `NULLABILITY_NULLABLE`.

```json
{
  "selection": {
    "directReference": {
      "structField": {
        "field": 1
      }
    },
    "rootReference": {}
  }
}
```

Whereas SQL references field by names, Substrait always references fields 
numerically. This means that a Substrait expression only makes sense relative
to a certain schema. As we'll see later when we discuss relations, for a filter
relation this will be relative to the input schema, so the `1` here is referring 
to the second field of `products`.

!!! note
    Protobuf may not serialize fields with integer type and value 0, since 0 is the default.
    So if you instead saw `"structField": {}`, know that is is equivalent to 
    `"structField": { "field": 0 }`.

`"Computers"` will be translated to a literal expression:

```json
{
  "literal": {
    "string": "Computers"
  }
}
```

Both `IS NULL` and `INDEX_IN` will be scalar function expressions. Available 
functions in Substrait are defined in extension YAML files contained in 
https://github.com/substrait-io/substrait/tree/main/extensions. Additional
extensions may be created elsewhere. `IS NULL` is defined as a `is_null` function
in 

> [functions_comparison.yaml](https://github.com/substrait-io/substrait/blob/main/extensions/functions_comparison.yaml#:~:text=%2D-,name%3A%20%22is_null%22,-description%3A%20Whether)
> and `INDEX_IN` is defined as `index_in` function in
> [functions_set.yaml](https://github.com/substrait-io/substrait/blob/main/extensions/functions_set.yaml#:~:text=%2D-,name%3A%20%22index_in%22,-description%3A%20%3E).

Reviewer 2: ```suggestion
[functions_comparison.yaml](https://github.com/substrait-io/substrait/blob/main/extensions/functions_comparison.yaml)
and `INDEX_IN` is defined as `index_in` function in
[functions_set.yaml](https://github.com/substrait-io/substrait/blob/main/extensions/functions_set.yaml).
```
Minor nit: I might be missing something but is the hash part of these URLs essential?

Author: That extra part is to highlight the relevant text. It means it will always scroll to the latest version of the function, regardless of where it is in the file.

First, the expression for `INDEX_IN("Computers", categories)` is:

<!-- TODO: update once #342 is merged -->

```json
{
  "scalarFunction": {
    "functionReference": 1,
    "outputType": {
      "i64": {
        "nullability": "NULLABILITY_NULLABLE"
      }
    },
    "arguments": [
      {
        "enum": {
          "specified": "NAN_IS_NOT_NAN"
        }
      },
      {
        "value": {
          "literal": {
            "string": "Computers"
          }
        }
      },
      {
        "value": {
          "selection": {
            "directReference": {
              "structField": {
                "field": 1
              }
            },
            "rootReference": {}
          }
        }
      }
    ]
  }
}
```

`functionReference` will be explained later in the plans section. For now,

> understand that it's an index into a list of function definitions and since this

Reviewer 2: Let's keep this unless you can come up with something better but this is technically not an "index into a list".  A functionReference of 7 could be the first item in the list or the 10th item.  It's more a key into a map.  However, I can't think of better wording without confusing a reader.

Author: I rephrased as an "ID corresponding to an entry in a list" which is hopefully clear and more accurate.

`outputType` defines the type the function outputs. We know this is a nullable
`i64` type since that is what the function definition declares in the YAML file.

`arguments` defines the arguments being passed into the function, which are all
done positionally based on the function definition in the YAML file. The first 
argument is an option that defines how `NaN` equality is handled, and since
that's not relevant here we just pass `"NAN_IS_NOT_NAN"`. The next two options
will be familiar as the literal and the field reference we constructed earlier.

To create the final expression, we just need to wrap this in another scalar
function expression for `IS NULL`.

```json
{
  "scalarFunction": {
    "functionReference": 2,
    "outputType": {
      "bool": {
        "nullability": "NULLABILITY_REQUIRED"
      }
    },
    "arguments": [
      {
        "value": {
          "scalarFunction": {
            "functionReference": 1,
            "outputType": {
              "i64": {
                "nullability": "NULLABILITY_NULLABLE"
              }
            },
            "arguments": [
              {
                "enum": {

>                   "specified": "NAN_IS_NOT_NAN"

Reviewer 2: Sorry, but this is out of date (my fault for the breaking change).  Options are no longer treated as arguments but rather their own thing.  So the `NAN_IS_NOT_NAN` will not be set in this way.

Author: Yup. Removed those references since they aren't required.

                              "field": 3
                            }
                          },
                          "rootReference": {}
                        }
                      }
                    }
                  ]
                }
              }
            }
          ]
        }
      }
    ]
  }
}
```

## Field indices

So far, we have glossed over the field indices. Now that we've built up each
of the relations, it will be a bit easier to explain them.

Throughout the plan, data always has some implicit schema, which is modified
by each relation. Often, the schema can change *within* a relation--we'll discuss
an example in the next section. Each relation has it's own rules in how schemas
are modified. For the purposes of our query, the relevant rules are:

* For Read relations, their output schema is the schema of the table.
* For Filter relations, the output schema is the same as in the input schema.
* For Joins relations, the input schema is the concatenation of the left and then the
  right schemas. The output schema is the same.
* For Aggregate relations, the output schema is the group by fields followed by
  the measures.

> <!-- TODO: do we document these rules somewhere? -->

Reviewer 2: In the docs / spec this is referred to as the "direct output order"

![image](https://user-images.githubusercontent.com/1696093/206299475-17a02cc2-3f57-47e9-9e3d-bf027c62da6b.png)

Author: Also seems like "emit order" is used. I've added mention of those two phrases to the doc so readers will recognize them.
